### ISSUE TITLE
Problem with the results

### ISSUE THREAD

**Issue by Anitha-byt** [ArnovanHilten/GenNet#89]
Hello Arno,

I am trying to implement the GENET -classification model for obesity.  I prepared the input files as mentioned in the instructions and I trained the model with default hyperparameters and also I tried to tune them but  I don't see good results and my model is randomly guessing. The validation loss seems to be fluctuating and does not seem to decrease gradually and  low sensitivity . I have more false positives and false negatives in the confusion matrix. Could you please suggest anything with regard to this?

**Comment by ArnovanHilten** [comment]
Hi Anitha,

There are a couple of hyperparameters you might need to tune. Most importantly the learning rate and the L1 penalty. If you post the learning curve I can provide some clues in which direction you should move but in the end this is often a bit of trial-and-error. 

If the network predicts only one class you can use the -wpc (weight for the positive class) input to increase or decrease the penalty for misclassifying the positive class. By default this is based on the ratio of the classe, the imbalance, in the training set but in my experience it is sometimes necessary to tune this.

There can be several other things that can be important. For example the number of SNPs and the number of training examples. If you provide the learning curve and some specifics of your training set I might be able to provide better advice.

Best,

Arno

**Comment by Anitha-byt** [comment]
Hi Arno, 
The Data has 8306 controls and 8306 cases. The no. of SNPs=131503. I have kept the default settings with lr=0.001, batch size=32 and and no. of epochs=50. I have attached the results.
Please let me know opinion


[model_architecture.txt](https://github.com/ArnovanHilten/GenNet/files/12362200/model_architecture.txt)
[train_log.csv](https://github.com/ArnovanHilten/GenNet/files/12362204/train_log.csv)
![train_val_loss](https://github.com/ArnovanHilten/GenNet/assets/83635519/10d1e4be-e768-4e43-9049-f7cc24377a4d)
[pd_summary_results.csv](https://github.com/ArnovanHilten/GenNet/files/12362205/pd_summary_results.csv)
[results_summary.txt](https://github.com/ArnovanHilten/GenNet/files/12362206/results_summary.txt)

**Comment by ArnovanHilten** [comment]
Hmmm it is a bit weird that the train loss is increasing. The default values that I provide are the parameters I found worked best for my applications but like any neural network you will need to do a hyperparameter search. Based on the learning curve I would suggest a smaller learning rate. You generally want a smoothly decreasing learning rate for the training like the picture below.

![image](https://github.com/ArnovanHilten/GenNet/assets/25904977/078f35b2-9d6e-47e8-887f-ae0acbf08d88)

Please let me know if the train loss keeps increasing with lower learning rates.

**Comment by Anitha-byt** [comment]
ok. Thank you so much. I will give it a try

**Comment by ArnovanHilten** [comment]
Hi Anitha, 

Yes, if you use the topology.csv you will only get networks that have a single gene layer. If you put multiple mask_name.npz in the folder you will get networks that have as many locallydirected layers as there are .npz files. Note: if your first npz has the shape [131503, 17825] than the second should have a shape of [178525, ?]. If you want to make some more specific changes you can also design your own network with this example: https://github.com/ArnovanHilten/GenNet/blob/bd92d2b41bcea8053adaf7ec5a53148e68f3205f/GenNet_utils/Create_network.py#L187 and add the new network function to the train_network.py.

I think adding more layers makes sense in your case since the train loss is not reaching 0. A good network optimization step is to try to get a zero loss for your train and then add regularisation to optimize for your val loss.

I have also added the option to use more "filters" akin to conv2d filters (see gene_network_multiple_filters in Create_network.py, use -network_name gene_network_multiple_filters and use .npz instead of topology).So each gene node is duplicated and merged with a subsequent locallyconnected1d layer. This increases the 'depth'/capabilities of the network in terms of filters instead of adding extra layers.

Seems that you are heading in the right direction. Good luck with your project!

Best,

Arno

**Comment by Anitha-byt** [comment]
Thanks again!

**Comment by Anitha-byt** [comment]
Hello again!,

I used -create_pathway_KEGG to generate the Masks. I see my network architecture as below but shape are not matching as you said. How do I make them same? Does the architecture seem ok?
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_layer (InputLayer)        [(None, 131503)]     0                                            
__________________________________________________________________________________________________
reshape (Reshape)               (None, 131503, 1)    0           input_layer[0][0]                
__________________________________________________________________________________________________
LocallyDirected_0 (LocallyDirec (None, 17825, 1)     133751      reshape[0][0]                    
__________________________________________________________________________________________________
activation (Activation)         (None, 17825, 1)     0           LocallyDirected_0[0][0]          
__________________________________________________________________________________________________
batch_normalization (BatchNorma (None, 17825, 1)     2           activation[0][0]                 
__________________________________________________________________________________________________
LocallyDirected_1 (LocallyDirec (None, 321, 1)       148433      batch_normalization[0][0]        
__________________________________________________________________________________________________
activation_1 (Activation)       (None, 321, 1)       0           LocallyDirected_1[0][0]          
__________________________________________________________________________________________________
batch_normalization_1 (BatchNor (None, 321, 1)       2           activation_1[0][0]               
__________________________________________________________________________________________________
flatten (Flatten)               (None, 321)          0           batch_normalization_1[0][0]      
__________________________________________________________________________________________________
output_layer (Dense)            (None, 1)            322         flatten[0][0]                    
__________________________________________________________________________________________________
activation_2 (Activation)       (None, 1)            0           output_layer[0][0]               
__________________________________________________________________________________________________
inputs_cov (InputLayer)         [(None, 9)]          0                                            
__________________________________________________________________________________________________
concatenate (Concatenate)       (None, 10)           0           activation_2[0][0]               
                                                                 inputs_cov[0][0]                 
__________________________________________________________________________________________________
batch_normalization_2 (BatchNor (None, 10)           20          concatenate[0][0]                
__________________________________________________________________________________________________
output_layer_cov (Dense)        (None, 1)            11          batch_normalization_2[0][0]      
__________________________________________________________________________________________________
activation_3 (Activation)       (None, 1)            0           output_layer_cov[0][0]           
==================================================================================================
Total params: 282,541
Trainable params: 282,517
Non-trainable params: 24

**Comment by ArnovanHilten** [comment]
Well depends on what you intent to create. You seen to have a gene layer with 17825 nodes and the first local pathway layer of KEGG (did you just use a single .npz file gene_local.npz? I think you are missing some higher level pathway masks if you wanted to make the standard pathway network (look for something like local_mid.npz and mid_global.npz the -create_pathway_KEGG should make them). 

If you just wanted to add a single layer it seems you succeeded.

You also added 9 covariates so that is added as extra input at the end but I would not consider that a full layer since it adds only 11 learnable parameters.

**Comment by Anitha-byt** [comment]
yes. I used only the first two masks. The validation becomes constant after some 5-10 epochs. I am trying to experiment with different learning rate values and  L1 regularisation.
